spec/awful/completion_spec.lua

```diff
@@ -0,0 +1,52 @@
+local compl = require("awful.completion")
+local shell = function(...)
+    local command, pos, matches = compl.shell(...)
+    return {command, pos, matches}
+end
+local gfs = require("gears.filesystem")
+
+describe("awful.completion.shell in empty directory", function()
+    local orig_popen = io.popen
+
+    setup(function()
+        gfs.mkdir('/tmp/awesome-tests/empty')
+        io.popen = function(...)
```

Comment:
The following causes no warnings here:
```lua
io.popen = function() end --luacheck: ignore
```
But since you are asking for "properly": You could do `completion.popen = io.popen` in the module and use that instead of `popen`. That way the tests could overwrite this directly instead of messing with the `io` table. (But I think this module-field should not be documented)